File: 1345-jump-game-iv/solution.py
from collections import defaultdict, deque

class Solution:
    def minJumps(self, arr: list[int]) -> int:
        n = len(arr)
        if n <= 1:
            return 0
        
        # Map values to their indices
        indices_of_value = defaultdict(list)
        for i, val in enumerate(arr):
            indices_of_value[val].append(i)
            
        queue = deque([0])
        visited = {0}
        steps = 0
        
        while queue:
            # Process current level
            for _ in range(len(queue)):
                i = queue.popleft()
                
                # Check if we reached the end
                if i == n - 1:
                    return steps
                
                # Potential next indices
                next_indices = [i - 1, i + 1]
                
                # Add jumps to indices with same value
                if arr[i] in indices_of_value:
                    next_indices.extend(indices_of_value[arr[i]])
                    # Important: Clear the entry to prevent redundant checks
                    # and ensure O(N) time complexity.
                    del indices_of_value[arr[i]]
                
                for next_i in next_indices:
                    if 0 <= next_i < n and next_i not in visited:
                        visited.add(next_i)
                        queue.append(next_i)
            
            steps += 1
            
        return -1

File: 1345-jump-game-iv/solution.cpp
#include <vector>
#include <unordered_map>
#include <queue>

using namespace std;

class Solution {
public:
    int minJumps(vector<int>& arr) {
        int n = arr.size();
        if (n <= 1) return 0;

        // Map values to all indices containing that value
        unordered_map<int, vector<int>> indicesOfValue;
        for (int i = 0; i < n; i++) {
            indicesOfValue[arr[i]].push_back(i);
        }

        queue<int> q;
        vector<bool> visited(n, false);

        q.push(0);
        visited[0] = true;
        int steps = 0;

        while (!q.empty()) {
            int size = q.size();
            while (size--) {
                int i = q.front();
                q.pop();

                // Target reached
                if (i == n - 1) return steps;

                // Move 1: i + 1
                if (i + 1 < n && !visited[i + 1]) {
                    visited[i + 1] = true;
                    q.push(i + 1);
                }

                // Move 2: i - 1
                if (i - 1 >= 0 && !visited[i - 1]) {
                    visited[i - 1] = true;
                    q.push(i - 1);
                }

                // Move 3: jump to indices with same value
                if (indicesOfValue.count(arr[i])) {
                    for (int nextIndex : indicesOfValue[arr[i]]) {
                        if (!visited[nextIndex]) {
                            visited[nextIndex] = true;
                            q.push(nextIndex);
                        }
                    }
                    // Optimization: Once all same-value jumps are processed, 
                    // clear the list to avoid O(N^2) behavior.
                    indicesOfValue.erase(arr[i]);
                }
            }
            steps++;
        }

        return -1;
    }
};